spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,223 @@
+---
+ics: 29
+title: General Fee Payment
+stage: draft
+category: IBC/APP
+requires: 20, 25, 26
+kind: instantiation
+author: Ethan Frey <ethan@confio.tech>
+created: 2021-06-01
+modified: 2021-06-01
+---
+
+## Synopsis
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for handling fee
+payments on top of any ICS application protocol. It requires some standard packet changes, but can be adopted by any
+application, without forcing other applications to use this implementation.
+
+### Motivation
+
+There has been much discussion on a general incentivization mechanism for relayers. A simple proposal was created to
+[extend ICS20 to incentivize relaying](https://github.com/cosmos/ibc/pull/577) on the destination chain. However,
+it was very specific to ICS20 and would not work for other protocols. This was then extended to a more
+[general fee payment design](https://github.com/cosmos/ibc/issues/578) that could be adopted by any ICS application
+protocol.
+
+In general, the Interchain dream will never scale unless there is a clear way to incentivize relayers. We seek to
+define a clear interface that can be easily adopted by any application, but not preclude chains that don't use tokens.
+
+### Desired Properties
+
+- Incentivize timely delivery of the packet (`OnReceivePacket` called)
+- Incentivize relaying acks for these packets (`OnAcknowledgement` called)
+- Incentivize relaying timeouts for these packets when the timeout has expired before packet is delivered (for example as receive fee was too low) (`OnTimeout` called)
+- Produces no extra IBC packets
+- One direction works, even when one chain does not support concept of fungible tokens
+- Opt-in for each chain implementing this. eg. ICS27 with fee support on chain A could connect to ICS27 without fee support on chain B.
+- Standardized interface for each chain implementing this extension
+- Support custom fee-handling logic within the same framework
+- Relayer addresses should not be forgable
+- Permissionless relaying
+
```

Comment:
Are we considering potential race conditions between relayers (i.e. where multiple relayers try to submit a packet in order to claim the fee and most of their transactions fail)? This is already a bit of a concern with IBC, but certain ways of paying fees would probably make it much worse, I think we should keep this in mind.

Reply:
Yes we've looked at permissionless relaying since that seems easiest from legal compliance standpoint, but I agree with you.

I also think there was a comment by @zmanian about allowing any altruistic relayer to relay packet, but only rewarding fee if it is a specific relayer potentially.

I've changed the interfaces to support "permissioned" relaying, so that if it is legal we can trivially support it from a technical standpoint

cc: @charleenfei

Reply:
> I don't think running relayers will be a profitable enough business (at this point and volume of ibc tx) for a hugely competitive relayer market, so really the incentivisation is to allow for relayer operators to at least cover their costs and will be more of a way for projects to offset costs than for a independent business model at the moment.

I agree here.

If it does become hugely profitable at some point, then I would assume that it would become like DeFi front-running. And in the end the validators will run the relayers and just relay on the blocks they propose, putting their messages first. Which is actually quite efficient and still relatively decentralized.

I assume we would only have 2 or 3 relayers trying not to go broke. But I also think the end-game of fee competition will not be so horrible. 

I also agree with technically allowing optional permissioned relaying. But think users should not be able to set the "selected relayer" on their packets unless it is enabled by chain governance (the legal burden is not on the users)